python/ray/tune/api_v2/tuner.py

```diff
@@ -0,0 +1,231 @@
+                Type[ConvertibleToTrainable],
+                ConvertibleToTrainable,
+            ]
+        ] = None,
+        run_config: Optional[Dict[str, Any]] = None,
+        param_space: Optional[Dict[str, Any]] = None,
+        name: Optional[str] = None,
+        callbacks: Optional[List[Callback]] = None,
+    ):
+        # Start from restore
+        if restore_path:
+            trainable_ckpt = os.path.join(restore_path, "trainable.pkl")
+            with open(trainable_ckpt, "rb") as fp:
+                trainable = pickle.load(fp)
+
+            tuner_ckpt = os.path.join(restore_path, "tuner.pkl")
+            with open(tuner_ckpt, "rb") as fp:
+                tuner = pickle.load(fp)
+                self.__dict__.update(tuner.__dict__)
+
+            self.is_restored = True
+            self.trainable = trainable
+            self.experiment_path = restore_path
+            return
+
+        # Start from fresh
+        assert trainable
+        self.is_restored = False
+        self.trainable = trainable
+        self.run_config = run_config
+        # Only used when constructing Tuner from scatch.
+        # Not used for restored Tuner.
+        self.param_space = param_space
+        self._process_dataset_param()
+
+        self.name = name
+        self.callbacks = callbacks
+
+        self.experiment_path = self._get_experiment_checkpoint_dir()
+
+        # This needs to happen earlier before run() is kicked in.
+        tuner_ckpt = os.path.join(self.experiment_path, "tuner.pkl")
+        with open(tuner_ckpt, "wb") as fp:
+            pickle.dump(self, fp)
+
+        trainable_ckpt = os.path.join(self.experiment_path, "trainable.pkl")
+        with open(trainable_ckpt, "wb") as fp:
+            pickle.dump(self.trainable, fp)
+
+    def _process_dataset_param(self):
+        """Some processing on Dataset param."""
+
+        def _helper(dataset_dict: dict):
+            for k, v in dataset_dict.items():
+                if isinstance(v, dict):
+                    _helper(v)
+                elif isinstance(v, Dataset):
+                    dataset_dict[k] = v.fully_executed()
+                elif isinstance(v, int):
+                    # CV settings
+                    pass
+                elif isinstance(v, list):
+                    if not all([isinstance(v_item, int) for v_item in v]) and not all(
+                        [isinstance(v_item, Dataset) for v_item in v]
+                    ):
+                        raise TuneError(
+                            "Wrongly formatted dataset param passed in Tune!"
+                        )
+                    if len(v) > 0 and isinstance(v[0], Dataset):
+                        dataset_dict[k] = [v_item.fully_executed() for v_item in v]
+                else:
+                    # We shouldn't be expecting anything here.
+                    raise TuneError("Unexpected dataset param passed in.")
+
+        if DATASETS in self.param_space:
```

Comment:
Should we have general support for this in the future, i.e. check all parameters if they are datasets (not just `DATASETS`)? Or will this become unnecessary anyway after ray data fixes?